operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,646 @@
+
+	for _, downedNode := range health.NodesDown {
+		ordinal, ok := brokerMap[downedNode]
+		if !ok {
+			// skip because we can't actually determine whether we should
+			// decommission it or not without its ordinal
+			brokersToIgnore = append(brokersToIgnore, downedNode)
+		}
+
+		status, err := adminClient.DecommissionBrokerStatus(ctx, downedNode)
+		if err != nil {
+			if strings.Contains(err.Error(), "is not decommissioning") {
+				if ordinal >= requestedNodes {
+					// this broker is old and should be deleted
+					brokersToDecommission = append(brokersToDecommission, downedNode)
+					continue
+				}
+
+				brokers := brokerOrdinalMap[ordinal]
+				if brokers.Size() == 1 {
+					// just ignore the node since it may be down, but it probably
+					// is just having problems
+					brokersToIgnore = append(brokersToIgnore, downedNode)
+					continue
+				}
+
+				// here we have multiple ordinals that align to different nodes
+				// and we're within our set ordinal range, make sure at least one
+				// other node in the set is healthy and then we can mark this
+				// node for decommission, otherwise, we can't distinguish which
+				// pod is which broker (i.e. they're all down) and whether we
+				// should actually decommission it or not
+				hasHealthyBroker := false
+				for _, broker := range brokers.Values() {
+					if broker == downedNode {
+						continue
+					}
+					if healthyNodes.HasAny(broker) {
+						hasHealthyBroker = true
+						break
+					}
+				}
+
+				if hasHealthyBroker {
+					// we have a healthy broker that isn't us, we can mark this for decommissioning
+					brokersToDecommission = append(brokersToDecommission, downedNode)
+					continue
+				}
+
+				// we can't tell which broker mapped to an ordinal is the current broker that
+				// may actually correspond to a still-existing pod, so just ignore this broker
+				brokersToIgnore = append(brokersToIgnore, downedNode)
+				continue
+			}
+
+			if strings.Contains(err.Error(), "does not exist") {
+				// delete the node from our sets
+				downedNodes.Delete(downedNode)
+				continue
+			}
+
+			log.Error(err, "fetching decommission status")
+			return false, err
+		}
+
+		if status.Finished {
+			// skip since we have already decommissioned it, so it should no longer
+			// show up in the health overview
+			continue
+		}
+
+		// add the brokers to the list of what needs to be decommissioned
+		currentlyDecommissioningBrokers = append(currentlyDecommissioningBrokers, downedNode)
+	}
+
+	sortBrokers := func(set []int) []int {
+		// sort by simple node id
+		sort.SliceStable(set, func(i, j int) bool {
+			return set[i] < set[j]
+		})
+		return set
+	}
+
+	healthyBrokers := sortBrokers(healthyNodes.Values())
+	brokersToDecommission = sortBrokers(brokersToDecommission)
+	brokersToIgnore = sortBrokers(brokersToIgnore)
+	currentlyDecommissioningBrokers = sortBrokers(currentlyDecommissioningBrokers)
+
+	formatBrokerList := func(set []int) string {
```

Comment:
just to be a pedant why not just use `%v`, `%#v`, or `%+v` to format?

Reply:
for a pedantic response 😛

**Example:**

 `fmt.Printf(..., []int{1, 2, 3, 4})`:

```
%v - [1 2 3 4]
%+v - [1 2 3 4]
%#v -[]int{1, 2, 3, 4}
```

I like commas without type information for readability, my format:

```
[1, 2, 3, 4]
```

Reply:
I appreciate the thoroughness 😄